Firestore/core/src/firebase/firestore/api/document_reference.h

```diff
@@ -50,14 +51,16 @@ class DocumentReference {
   using Completion = void (^)(NSError* _Nullable error) _Nullable;
 
   DocumentReference() = default;
-  DocumentReference(model::ResourcePath path, Firestore* firestore);
-  DocumentReference(model::DocumentKey document_key, Firestore* firestore)
+  DocumentReference(model::ResourcePath path,
+                    std::shared_ptr<Firestore> firestore);
+  DocumentReference(model::DocumentKey document_key,
+                    std::shared_ptr<Firestore> firestore)
       : firestore_{firestore}, key_{std::move(document_key)} {
   }
 
   size_t Hash() const;
 
-  Firestore* firestore() const {
+  const std::shared_ptr<Firestore>& firestore() const {
```

Comment:
Did you run into any problems with this reference being marked `const`?

Intuitively to me this seems as if this makes the returned `Firestore` pointer strictly less useful than what came before because very few methods on the `Firestore` class are `const`.

However, I guess this is working because the callers are generally copying this into a non-const `std::shared_ptr<Firestore>`.

Reply:
This (as I understand it) is just making the `shared_ptr` const, which e.g. means the caller can't call `reset()` on it.  The `Firestore` being pointed to is not const.  For that, I think it'd be `shared_ptr<const Firestore>`...

Reply:
That's not quite right: the shared_ptr is copied to the caller, so the caller can still reset it. The (first) const really doesn't do much here. eg: this works:
```c++
#include <memory>
  
struct Foo {};

const std::shared_ptr<Foo> get_const_foo() {
  return std::make_shared<Foo>();
}
  
int main() {
  std::shared_ptr<Foo> f = get_const_foo();
  f.reset();
}
```

Reply:
I think the problem comes up if you use it as an r-value (avoiding the copy).  i.e.: this:

```
get_const_foo().reset();
```

[gives you the error](https://onlinegdb.com/rypxAEf5V):

```
main.cpp:18:25: error: no matching function for call to ‘std::shared_ptr::reset() const’
   get_const_foo().reset();
```

So although it's unlikely somebody would do this, I think `const` is beneficial here to prevent mutation of the shared_ptr.

Reply:
Making the return type a const reference to the shared_ptr makes the exposed Firestore instance available without bumping the reference count. If the caller just uses it as an r-value (i.e. to indirect through the pointer, this is cheap).

Returning the shared_ptr by value forces a copy (and thus bumps the reference count). This makes this less cheap. In @rsgowman's example, the caller is resetting its copy of the shared_ptr. If `get_const_foo()` were returning a member variable, the member would be unaffected.

Given this understanding I think the const reference to shared_ptr as a return type here is strictly better.